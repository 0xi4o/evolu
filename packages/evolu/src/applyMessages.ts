import { option, readonlyArray, taskEither } from "fp-ts";
import { flow, pipe } from "fp-ts/lib/function.js";
import { ReaderTaskEither } from "fp-ts/ReaderTaskEither";
import { ReadonlyNonEmptyArray } from "fp-ts/ReadonlyNonEmptyArray";
import { insertIntoMerkleTree } from "./merkleTree.js";
import { timestampFromString } from "./timestamp.js";
import {
  CrdtMessage,
  DbEnv,
  MerkleTree,
  TimestampString,
  UnknownError,
} from "./types.js";

export const applyMessages =
  (merkleTree: MerkleTree) =>
  (
    messages: ReadonlyNonEmptyArray<CrdtMessage>
  ): ReaderTaskEither<DbEnv, UnknownError, MerkleTree> =>
  ({ db }) =>
    pipe(
      messages,
      taskEither.traverseSeqArray((message) =>
        pipe(
          db.execSqlQuery({
            sql: `
              select "timestamp" FROM "__message"
              where "table" = ? AND
                    "row" = ? AND
                    "column" = ?
              order by "timestamp" desc limit 1
            `,
            parameters: [message.table, message.row, message.column],
          }),
          taskEither.map(
            flow(
              readonlyArray.head,
              option.map((row) => row.timestamp as TimestampString),
              option.toNullable
            )
          ),
          taskEither.chainFirst((timestamp) =>
            timestamp == null || timestamp < message.timestamp
              ? db.execSqlQuery({
                  sql: `
                    insert into "${message.table}" ("id", "${message.column}")
                    values (?, ?)
                    on conflict do update set "${message.column}" = ?
                  `,
                  parameters: [message.row, message.value, message.value],
                })
              : taskEither.right(undefined)
          ),
          taskEither.chainFirst((timestamp) =>
            timestamp == null || timestamp !== message.timestamp
              ? pipe(
                  db.execSqlQuery({
                    sql: `
                      insert into "__message" (
                        "timestamp", "table", "row", "column", "value"
                      ) values (?, ?, ?, ?, ?) on conflict do nothing
                    `,
                    parameters: [
                      message.timestamp,
                      message.table,
                      message.row,
                      message.column,
                      message.value,
                    ],
                  }),
                  taskEither.chain(db.changes),
                  taskEither.map((changes) => {
                    if (changes > 0)
                      merkleTree = insertIntoMerkleTree(
                        timestampFromString(message.timestamp)
                      )(merkleTree);
                  })
                )
              : taskEither.right(undefined)
          )
        )
      ),
      taskEither.map(() => merkleTree)
    );
