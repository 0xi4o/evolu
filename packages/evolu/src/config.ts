import { Config } from "./types.js";

export let config: Config = {
  syncUrl: "https://bold-frost-4029.fly.dev",
  log: false,
  maxDrift: 60000,
  reloadUrl: "/",
};

export const setConfig = (c: Config): void => {
  config = c;
};
