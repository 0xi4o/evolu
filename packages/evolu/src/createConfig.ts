import { IO } from "fp-ts/lib/IO.js";
import { Config } from "./types.js";

const defaultConfig: Config = {
  syncUrl: "https://bold-frost-4029.fly.dev",
  maxDrift: 60000,
  reloadUrl: "/",
};

export const createConfig =
  (config?: Partial<Config>): IO<Config> =>
  () => ({ ...defaultConfig, ...config });
