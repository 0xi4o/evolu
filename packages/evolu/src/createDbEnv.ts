import { either, taskEither } from "fp-ts";
import { pipe } from "fp-ts/lib/function.js";
import { TaskEither } from "fp-ts/TaskEither";
import sqlite3 from "../sqlite/sqlite3-bundler-friendly.mjs";
import {
  Database,
  DbEnv,
  errorToUnknownError,
  SqliteRows,
  UnknownError,
} from "./types.js";

export const createDbEnv: TaskEither<UnknownError, DbEnv> = taskEither.tryCatch(
  () =>
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    sqlite3().then((sqlite3: any): DbEnv => {
      const db = new sqlite3.oo1.OpfsDb("/evolu/evolu1.db", "c");

      const exec: Database["exec"] = (sql) =>
        pipe(
          either.tryCatch(
            (): SqliteRows =>
              db.exec(sql, {
                returnValue: "resultRows",
                rowMode: "object",
              }),
            errorToUnknownError
          ),
          taskEither.fromEither
        );

      const execSqlQuery: Database["execSqlQuery"] = (sqlQuery) =>
        pipe(
          either.tryCatch(
            () =>
              db.exec(sqlQuery.sql, {
                returnValue: "resultRows",
                rowMode: "object",
                bind: sqlQuery.parameters,
              }),
            errorToUnknownError
          ),
          taskEither.fromEither
        );

      const changes: Database["changes"] = () =>
        pipe(
          either.tryCatch(() => db.changes(), errorToUnknownError),
          taskEither.fromEither
        );

      return {
        db: {
          SQLite3Error: sqlite3.SQLite3Error,
          exec,
          execSqlQuery,
          changes,
        },
      };
    }),
  errorToUnknownError
);
