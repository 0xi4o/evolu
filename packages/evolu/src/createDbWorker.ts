import { IO } from "fp-ts/IO";
import { constVoid } from "fp-ts/lib/function.js";
import { CreateDbWorker, DbWorkerOutput, PostDbWorkerInput } from "./types.js";
import { isServer } from "./utils.js";

const isChromeWithOpfs: IO<boolean> = () =>
  navigator.userAgentData != null &&
  navigator.userAgentData.brands.find(
    ({ brand, version }) =>
      // Chrome or Chromium
      brand.includes("Chrom") && Number(version) >= 109
  ) != null;

const createNoOpServerDbWorker: CreateDbWorker = () => ({
  post: () => constVoid,
});

const createOpfsDbWorker: CreateDbWorker = (onMessage) => {
  const dbWorker = new Worker(new URL("./opfs.worker.js", import.meta.url));

  const post: PostDbWorkerInput = (message) => () =>
    dbWorker.postMessage(message);

  dbWorker.onmessage = (e: MessageEvent<DbWorkerOutput>): void =>
    onMessage(e.data)();

  return { post };
};

// TODO: LocalStorage, React Native, Electron.
export const createDbWorker: CreateDbWorker = isServer
  ? createNoOpServerDbWorker
  : isChromeWithOpfs()
  ? createOpfsDbWorker
  : createNoOpServerDbWorker;
