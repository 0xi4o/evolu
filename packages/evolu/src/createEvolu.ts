import {
  io,
  ioOption,
  ioRef,
  option,
  readonlyArray,
  readonlyNonEmptyArray,
  readonlyRecord,
  taskEither,
} from "fp-ts";
import { IO } from "fp-ts/IO";
import {
  constVoid,
  decrement,
  flow,
  increment,
  pipe,
} from "fp-ts/lib/function.js";
import { Option } from "fp-ts/Option";
import { ReadonlyNonEmptyArray } from "fp-ts/ReadonlyNonEmptyArray";
import { ReadonlyRecord } from "fp-ts/ReadonlyRecord";
import { flushSync } from "react-dom";
import { createStore } from "./createStore.js";
import { dbSchemaToTableDefinitions } from "./dbSchemaToTableDefinitions.js";
import { applyPatches } from "./diff.js";
import { cast, createId, CreateId, ID, SqliteDateTime } from "./model.js";
import { reloadAllTabs } from "./reloadAllTabs.js";
import {
  CreateEvolu,
  DbSchema,
  DbWorker,
  DbWorkerOutputOnQuery,
  eqSqlQueryString,
  Evolu,
  EvoluError,
  Mutate,
  NewCrdtMessage,
  OnCompleteId,
  Owner,
  OwnerActions,
  RowsCache,
  SqlQueryString,
  Store,
} from "./types.js";
import { isServer } from "./utils.js";

const createNewCrdtMessages = (
  table: string,
  row: ID<"string">,
  values: ReadonlyRecord<string, unknown>,
  ownerId: ID<"owner">,
  now: SqliteDateTime,
  isInsert: boolean
): ReadonlyNonEmptyArray<NewCrdtMessage> =>
  pipe(
    readonlyRecord.toEntries(values),
    readonlyArray.filter(([, value]) => value !== undefined),
    readonlyArray.map(([key, value]) => [
      key,
      typeof value === "boolean" || value instanceof Date
        ? cast(value as never)
        : value,
    ]),
    isInsert
      ? flow(
          readonlyArray.appendW(["createdAt", now]),
          readonlyArray.appendW(["createdBy", ownerId])
        )
      : readonlyArray.appendW(["updatedAt", now]),
    readonlyNonEmptyArray.map(
      ([column, value]) =>
        ({
          table,
          row,
          column,
          value,
        } as NewCrdtMessage)
    )
  );

const createMutate = <S extends DbSchema>({
  createId,
  getOwner,
  setOnComplete,
  dbWorker,
  getSubscribedQueries,
}: {
  createId: CreateId;
  getOwner: Promise<Owner>;
  setOnComplete: (id: OnCompleteId, callback: IO<void>) => void;
  dbWorker: DbWorker;
  getSubscribedQueries: IO<readonly SqlQueryString[]>;
}): Mutate<S> => {
  const mutateQueueRef = new ioRef.IORef<
    readonly {
      readonly messages: ReadonlyNonEmptyArray<NewCrdtMessage>;
      readonly onCompleteId: Option<OnCompleteId>;
    }[]
  >(readonlyArray.empty);

  return (table, { id, ...values }, onComplete) => {
    const isInsert = id == null;
    if (isInsert) id = createId() as never;
    const now = cast(new Date());

    getOwner.then((owner) => {
      const messages = createNewCrdtMessages(
        table as string,
        id as ID<"string">,
        values,
        owner.id,
        now,
        isInsert
      );

      const onCompleteId = pipe(
        onComplete,
        option.fromNullable,
        option.map((onComplete) => {
          const id: OnCompleteId = createId<"OnComplete">();
          setOnComplete(id, onComplete);
          return id;
        })
      );

      const runQueueMicrotask = mutateQueueRef.read().length === 0;
      mutateQueueRef.modify(readonlyArray.append({ messages, onCompleteId }))();

      if (runQueueMicrotask)
        pipe(
          mutateQueueRef.read,
          io.chainFirst(() => mutateQueueRef.write([])),
          io.map(readonlyNonEmptyArray.fromReadonlyArray),
          ioOption.chainIOK((queue) =>
            dbWorker.post({
              type: "send",
              messages: pipe(
                queue,
                readonlyNonEmptyArray.map((a) => a.messages),
                readonlyNonEmptyArray.flatten
              ),
              onCompleteIds: queue
                .map((a) => a.onCompleteId)
                .filter(option.isSome)
                .map((a) => a.value),
              queries: getSubscribedQueries(),
            })
          ),
          queueMicrotask
        );
    });

    return { id } as never;
  };
};

const createOnQuery =
  (rowsCache: Store<RowsCache>, onCompletes: Map<OnCompleteId, IO<void>>) =>
  ({ queriesPatches, onCompleteIds }: DbWorkerOutputOnQuery): IO<void> =>
  () => {
    pipe(
      queriesPatches,
      readonlyArray.reduce(
        rowsCache.getState(),
        (state, { query, patches }) => ({
          ...state,
          [query]: applyPatches(patches)(state[query]),
        })
      ),
      (state) => {
        if (onCompleteIds.length === 0) {
          rowsCache.setState(state)();
          return;
        }

        // flushSync is required before callOnCompletes
        if (queriesPatches.length > 0)
          flushSync(() => {
            rowsCache.setState(state)();
          });

        pipe(
          onCompleteIds,
          readonlyArray.filterMap((id) => {
            const onComplete = onCompletes.get(id);
            onCompletes.delete(id);
            return option.fromNullable(onComplete);
          })
        ).forEach((onComplete) => onComplete());
      }
    );
  };

const createSubscribeQuery = (
  rowsCache: Store<RowsCache>,
  subscribedQueries: Map<SqlQueryString, number>,
  queryIfAny: (queries: readonly SqlQueryString[]) => IO<void>
): Evolu<never>["subscribeQuery"] => {
  const snapshot = new ioRef.IORef<readonly SqlQueryString[] | null>(null);

  return (sqlQueryString: SqlQueryString | null) => (listen) => {
    if (sqlQueryString == null) return () => constVoid;

    if (snapshot.read() == null) {
      snapshot.write(Array.from(subscribedQueries.keys()))();
      queueMicrotask(() => {
        const subscribedQueriesSnapshot = snapshot.read();
        if (subscribedQueriesSnapshot == null) return;
        snapshot.write(null)();
        pipe(
          Array.from(subscribedQueries.keys()),
          readonlyArray.difference(eqSqlQueryString)(subscribedQueriesSnapshot),
          queryIfAny
        )();
      });
    }

    subscribedQueries.set(
      sqlQueryString,
      increment(subscribedQueries.get(sqlQueryString) ?? 0)
    );
    const unsubscribe = rowsCache.subscribe(listen);

    return () => {
      const count = subscribedQueries.get(sqlQueryString);
      if (count != null && count > 1)
        subscribedQueries.set(sqlQueryString, decrement(count));
      else subscribedQueries.delete(sqlQueryString);
      unsubscribe();
    };
  };
};

const createOwnerActionRestore =
  (dbWorker: DbWorker): OwnerActions["restore"] =>
  (mnemonic) =>
    pipe(
      taskEither.fromTask(() => import("./mnemonic")),
      taskEither.chainEitherKW(({ parseMnemonic }) => parseMnemonic(mnemonic)),
      taskEither.chainIOK((mnemonic) =>
        dbWorker.post({ type: "restoreOwner", mnemonic })
      )
    );

const initReconnectAndReshow = (
  subscribedQueries: Map<SqlQueryString, number>,
  dbWorker: DbWorker
): void => {
  const sync = (refreshQueries: boolean): IO<void> =>
    pipe(
      () =>
        refreshQueries
          ? readonlyNonEmptyArray.fromArray(
              Array.from(subscribedQueries.keys())
            )
          : option.none,
      io.chain((queries) => dbWorker.post({ type: "sync", queries }))
    );

  const handleReconnect = sync(false);
  const handleReshow = sync(true);

  if (!isServer) {
    window.addEventListener("online", handleReconnect);
    window.addEventListener("focus", handleReshow);
    document.addEventListener("visibilitychange", () => {
      if (document.visibilityState !== "hidden") handleReshow();
    });
    handleReconnect();
  }
};

export const createEvolu: CreateEvolu =
  (dbSchema) =>
  ({ config, createDbWorker }) => {
    const errorStore = createStore<EvoluError | null>(null);
    const rowsCache = createStore<RowsCache>({});
    const ownerStore = createStore<Owner | null>(null);
    const onCompletes = new Map<OnCompleteId, IO<void>>();
    const subscribedQueries = new Map<SqlQueryString, number>();

    const onQuery = createOnQuery(rowsCache, onCompletes);

    const queryIfAny: (queries: readonly SqlQueryString[]) => IO<void> = flow(
      readonlyNonEmptyArray.fromReadonlyArray,
      option.match(
        () => io.of(constVoid),
        (queries) => dbWorker.post({ type: "query", queries })
      )
    );

    const dbWorker = createDbWorker((output) => {
      switch (output.type) {
        case "onError":
          return errorStore.setState({
            type: "EvoluError",
            error: output.error,
          });
        case "onOwner":
          return ownerStore.setState(output.owner);
        case "onQuery":
          return onQuery(output);
        case "onReceive":
          return queryIfAny(Array.from(subscribedQueries.keys()));
        case "onResetOrRestore":
          return reloadAllTabs(config.reloadUrl);
      }
    });

    const subscribeQuery = createSubscribeQuery(
      rowsCache,
      subscribedQueries,
      queryIfAny
    );

    const getQuery: Evolu<never>["getQuery"] = (query) => () =>
      (query && rowsCache.getState()[query]) || null;

    const getOwner = new Promise<Owner>((resolve) => {
      const unsubscribe = ownerStore.subscribe(() => {
        const o = ownerStore.getState();
        if (!o) return;
        unsubscribe();
        resolve(o);
      });
    });

    const mutate: Evolu<never>["mutate"] = createMutate({
      createId,
      getOwner,
      setOnComplete: (id, callback) => {
        onCompletes.set(id, callback);
      },
      dbWorker,
      getSubscribedQueries: () => Array.from(subscribedQueries.keys()),
    });

    const ownerActions: OwnerActions = {
      reset: dbWorker.post({ type: "resetOwner" }),
      restore: createOwnerActionRestore(dbWorker),
    };

    dbWorker.post({
      type: "init",
      config,
      tableDefinitions: dbSchemaToTableDefinitions(dbSchema),
    })();

    initReconnectAndReshow(subscribedQueries, dbWorker);

    return {
      subscribeError: errorStore.subscribe,
      getError: errorStore.getState,
      subscribeOwner: ownerStore.subscribe,
      getOwner: ownerStore.getState,
      subscribeQuery,
      getQuery,
      mutate,
      ownerActions,
    };
  };
