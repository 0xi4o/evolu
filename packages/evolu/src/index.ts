import * as S from "@effect/schema/Schema";
import { Effect, Function, Layer } from "effect";
import { Config, ConfigLive } from "./Config.js";
import { Schema, Tables, schemaToTables } from "./Db.js";
import { DbWorker, DbWorkerOutput } from "./DbWorker.js";
import { EvoluLive } from "./Evolu.js";
import { Platform } from "./Platform.js";
import { ReactHooks, ReactHooksLive } from "./React.js";
export * from "./exports.js";

import { Bip39Live, NanoIdLive } from "./CryptoLive.web.js";
import { AppStateLive, FlushSyncLive, PlatformLive } from "./Platform.web.js";
import { TimeLive } from "./Timestamp.js";

const DbWorkerLive = Layer.effect(
  DbWorker,
  Effect.gen(function* (_) {
    const platform = yield* _(Platform);

    if (platform.name === "web-with-opfs") {
      const worker = new Worker(
        new URL("DbWorker.worker.js", import.meta.url),
        { type: "module" },
      );
      worker.onmessage = (e: MessageEvent<DbWorkerOutput>): void => {
        dbWorker.onMessage(e.data);
      };
      const dbWorker: DbWorker = {
        postMessage: (input) => {
          worker.postMessage(input);
        },
        onMessage: Function.constVoid,
      };
      return dbWorker;
    }

    if (platform.name === "web-without-opfs") {
      const promise = Effect.promise(() => import("./DbWorker.web.js")).pipe(
        Effect.map(({ dbWorker: importedDbWorker }) => {
          importedDbWorker.onMessage = dbWorker.onMessage;
          return importedDbWorker.postMessage;
        }),
        Effect.runPromise,
      );
      const dbWorker: DbWorker = {
        postMessage: (input) => {
          void promise.then((postMessage) => {
            postMessage(input);
          });
        },
        onMessage: Function.constVoid,
      };
      return dbWorker;
    }

    return DbWorker.of({
      postMessage: Function.constVoid,
      onMessage: Function.constVoid,
    });
  }),
);

export const create = <From, To extends Schema>(
  schema: S.Schema<From, To>,
  config?: Partial<Config>,
): ReactHooks<To> => {
  return ReactHooks<To>().pipe(
    Effect.provideLayer(
      Layer.use(
        ReactHooksLive<To>(),
        Layer.merge(
          PlatformLive,
          Layer.use(
            EvoluLive<To>(),
            Layer.mergeAll(
              Layer.use(DbWorkerLive, PlatformLive),
              Bip39Live,
              ConfigLive(config),
              Layer.succeed(Tables, schemaToTables(schema)),
              FlushSyncLive,
              NanoIdLive,
              TimeLive,
              Layer.use(
                AppStateLive,
                Layer.merge(PlatformLive, ConfigLive(config)),
              ),
            ),
          ),
        ),
      ),
    ),
    Effect.runSync,
  );
};
