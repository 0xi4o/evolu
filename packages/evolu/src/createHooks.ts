import { Schema } from "@effect/schema";
import { readonlyArray } from "fp-ts";
import { constNull, pipe } from "fp-ts/lib/function.js";
import { useMemo, useRef, useSyncExternalStore } from "react";
import { createConfig } from "./createConfig.js";
import { createDbWorker } from "./createDbWorker.js";
import { createEvolu } from "./createEvolu.js";
import { kysely } from "./kysely.js";
import {
  Config,
  Create,
  DbSchema,
  Hooks,
  OrNullOrFalse,
  SqliteRow,
  SqlQuery,
  sqlQueryToString,
  Update,
  UseMutation,
  UseQuery,
} from "./types.js";

/**
 * Create React Hooks and initialize the database for a given schema.
 *
 * For domain modeling, Evolu uses [Schema](https://github.com/effect-ts/schema).
 *
 * Remember, you shall not remove any table or column after the first release.
 * There is a possibility that somebody is already using it. All you can do is
 * add new tables or columns and stop using obsolete ones. That's because it's
 * not feasible to migrate local data on any schema change.
 *
 * With this simple rule, any app version can handle any schema version.
 * If an obsolete app gets a sync message with a newer schema, Evolu
 * automatically updates the database schema to store the message safely.
 *
 * To learn more about handling schema evolving without migrations, check
 * the `useQuery` documentation.
 *
 * ### Example
 *
 * ```
 * import * as S from "@effect/schema";
 * import * as E from "evolu";
 *
 * const TodoId = E.id("Todo");
 * type TodoId = S.Infer<typeof TodoId>;
 *
 * const TodoTable = S.struct({
 *   id: TodoId,
 *   title: E.NonEmptyString1000,
 *   isCompleted: E.SqliteBoolean,
 * });
 * type TodoTable = S.Infer<typeof TodoTable>;
 *
 * const Database = S.struct({
 *   todo: TodoTable,
 * });
 *
 * const { useQuery, useMutation, useEvoluError, useOwner, useOwnerActions } =
 *   E.createHooks(Database);
 * ```
 */
export const createHooks = <S extends DbSchema>(
  dbSchema: Schema<S>,
  config?: Partial<Config>
): Hooks<S> => {
  const evolu = createEvolu(dbSchema)({
    config: createConfig(config),
    createDbWorker,
  });
  const cache = new WeakMap<SqliteRow, OrNullOrFalse<SqliteRow>>();

  const useQuery: UseQuery<S> = (query, filterMap) => {
    const sqlQueryString = query
      ? pipe(query(kysely as never).compile() as SqlQuery, sqlQueryToString)
      : null;

    const queryRows = useSyncExternalStore(
      useMemo(() => evolu.subscribeQuery(sqlQueryString), [sqlQueryString]),
      evolu.getQuery(sqlQueryString),
      constNull
    );

    const filterMapRef = useRef(filterMap);

    const getRowFromCache = (row: SqliteRow): OrNullOrFalse<SqliteRow> => {
      let cachedRow = cache.get(row);
      if (cachedRow !== undefined) return cachedRow;
      cachedRow = filterMapRef.current(row as never);
      cache.set(row, cachedRow);
      return cachedRow;
    };

    const rows = useMemo(() => {
      if (queryRows == null) return null;
      const rows: Array<SqliteRow> = [];
      queryRows.forEach((queryRow) => {
        const row = getRowFromCache(queryRow);
        if (row) rows.push(row);
      });
      return rows;
    }, [queryRows]);

    return useMemo(
      () => ({
        rows: (rows || readonlyArray.empty) as never,
        row: ((rows && rows[0]) || null) as never,
        isLoaded: rows != null,
      }),
      [rows]
    );
  };

  const useMutation: UseMutation<S> = () =>
    useMemo(
      () => ({
        create: evolu.mutate as Create<S>,
        update: evolu.mutate as Update<S>,
      }),
      []
    );

  const useEvoluError: Hooks<S>["useEvoluError"] = () =>
    useSyncExternalStore(evolu.subscribeError, evolu.getError, constNull);

  const useOwner: Hooks<S>["useOwner"] = () =>
    useSyncExternalStore(evolu.subscribeOwner, evolu.getOwner, constNull);

  const useOwnerActions: Hooks<S>["useOwnerActions"] = () => evolu.ownerActions;

  return {
    useQuery,
    useMutation,
    useEvoluError,
    useOwner,
    useOwnerActions,
  };
};
