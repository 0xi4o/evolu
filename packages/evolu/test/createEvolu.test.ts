import { expect, test } from "vitest";
import { Id, OwnerId, SqliteDate } from "../src";
import { createNewCrdtMessages } from "../src/createEvolu";

test("createNewCrdtMessages", () => {
  const values = {
    a: undefined,
    b: null,
    c: false,
    d: new Date(123),
    e: "string",
  };

  expect(
    createNewCrdtMessages(
      "table",
      "row" as Id,
      values,
      "ownerId" as OwnerId,
      "now" as SqliteDate,
      true
    )
  ).toMatchSnapshot();

  expect(
    createNewCrdtMessages(
      "table",
      "row" as Id,
      values,
      "ownerId" as OwnerId,
      "now" as SqliteDate,
      false
    )
  ).toMatchSnapshot();
});
