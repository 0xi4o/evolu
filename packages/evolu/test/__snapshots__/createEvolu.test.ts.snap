// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createNewCrdtMessages 1`] = `
[
  {
    "column": "c",
    "row": "row",
    "table": "table",
    "value": 0,
  },
  {
    "column": "d",
    "row": "row",
    "table": "table",
    "value": "1970-01-01T00:00:00.123Z",
  },
  {
    "column": "e",
    "row": "row",
    "table": "table",
    "value": "string",
  },
  {
    "column": "createdAt",
    "row": "row",
    "table": "table",
    "value": "now",
  },
  {
    "column": "createdBy",
    "row": "row",
    "table": "table",
    "value": "ownerId",
  },
]
`;

exports[`createNewCrdtMessages 2`] = `
[
  {
    "column": "b",
    "row": "row",
    "table": "table",
    "value": null,
  },
  {
    "column": "c",
    "row": "row",
    "table": "table",
    "value": 0,
  },
  {
    "column": "d",
    "row": "row",
    "table": "table",
    "value": "1970-01-01T00:00:00.123Z",
  },
  {
    "column": "e",
    "row": "row",
    "table": "table",
    "value": "string",
  },
  {
    "column": "updatedAt",
    "row": "row",
    "table": "table",
    "value": "now",
  },
]
`;
